From 6c05dc7563e61ab3c03cf8bdc87fd58bc01ac852 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Cl=C3=A9ment=20Le=20Goffic?= <clement.legoffic@foss.st.com>
Date: Mon, 2 Sep 2024 12:19:19 +0200
Subject: [PATCH 1145/1695] core: pta: add alarm-related operations to RTC PTA
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

- Added set_alarm and read_alarm operations.
- Introduced enable_alarm, wait_alarm, cancel_wait, and
  set_wake_alarm_status commands.
- Updated RTC features to include alarm and wakeup alarm capabilities.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 838146eeabf0e922ddacae80d18708a6d17d8c44 core: pta: add alarm-related operations to RTC PTA]
Change-Id: Ib8d7593e7b1069e544a7da4e9e3507d34668ca47
Signed-off-by: Clément Le Goffic <clement.legoffic@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/402359
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/pta/rtc.c                | 144 ++++++++++++++++++++++++++++++++++
 lib/libutee/include/pta_rtc.h |  98 +++++++++++++++++++++++
 2 files changed, 242 insertions(+)

diff --git a/core/pta/rtc.c b/core/pta/rtc.c
index ff69319d6..c2dd87733 100644
--- a/core/pta/rtc.c
+++ b/core/pta/rtc.c
@@ -39,6 +39,24 @@ static void rtc_pta_copy_time_to_optee(struct optee_rtc_time *optee_time,
 	optee_time->tm_wday = pta_time->tm_wday;
 }
 
+static void rtc_pta_copy_alarm_from_optee(struct pta_rtc_alarm *pta_alarm,
+					  struct optee_rtc_alarm *optee_alarm)
+{
+	pta_alarm->enabled = optee_alarm->enabled;
+	pta_alarm->pending = optee_alarm->pending;
+	rtc_pta_copy_time_from_optee(&pta_alarm->time,
+				     &optee_alarm->time);
+}
+
+static void rtc_pta_copy_alarm_to_optee(struct optee_rtc_alarm *optee_alarm,
+					struct pta_rtc_alarm *pta_alarm)
+{
+	optee_alarm->enabled = pta_alarm->enabled;
+	optee_alarm->pending = pta_alarm->pending;
+	rtc_pta_copy_time_to_optee(&optee_alarm->time,
+				   &pta_alarm->time);
+}
+
 static TEE_Result rtc_pta_get_time(uint32_t types,
 				   TEE_Param params[TEE_NUM_PARAMS])
 {
@@ -121,6 +139,116 @@ static TEE_Result rtc_pta_get_offset(uint32_t types,
 	return res;
 }
 
+static TEE_Result rtc_pta_read_alarm(uint32_t types,
+				     TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct optee_rtc_alarm alarm = { };
+	struct pta_rtc_alarm *pta_alarm = NULL;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	pta_alarm = params[0].memref.buffer;
+	if (!params[0].memref.buffer ||
+	    params[0].memref.size != sizeof(*pta_alarm) ||
+	    !IS_ALIGNED_WITH_TYPE(params[0].memref.buffer, typeof(*pta_alarm)))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = rtc_read_alarm(&alarm);
+	if (!res)
+		rtc_pta_copy_alarm_from_optee(pta_alarm, &alarm);
+
+	return res;
+}
+
+static TEE_Result rtc_pta_set_alarm(uint32_t types,
+				    TEE_Param params[TEE_NUM_PARAMS])
+{
+	struct optee_rtc_alarm alarm = { };
+	struct pta_rtc_alarm *pta_alarm = NULL;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	pta_alarm = params[0].memref.buffer;
+	if (!pta_alarm || params[0].memref.size != sizeof(*pta_alarm) ||
+	    !IS_ALIGNED_WITH_TYPE(pta_alarm, typeof(*pta_alarm)))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	rtc_pta_copy_alarm_to_optee(&alarm, pta_alarm);
+
+	return rtc_set_alarm(&alarm);
+}
+
+static TEE_Result rtc_pta_enable_alarm(uint32_t types,
+				       TEE_Param params[TEE_NUM_PARAMS])
+{
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (params[0].value.a != 0 && params[0].value.a != 1)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return rtc_enable_alarm((bool)params[0].value.a);
+}
+
+static TEE_Result rtc_pta_wait_alarm(uint32_t types,
+				     TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	enum rtc_wait_alarm_status return_status = RTC_WAIT_ALARM_RESET;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = rtc_wait_alarm(&return_status);
+	if (!res)
+		params[0].value.a = return_status;
+
+	return res;
+}
+
+static TEE_Result rtc_pta_cancel_wait(uint32_t types,
+				      TEE_Param params[TEE_NUM_PARAMS] __unused)
+{
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return rtc_alarm_cancel_wait();
+}
+
+static TEE_Result
+rtc_pta_alarm_wake_set_status(uint32_t types,
+			      TEE_Param params[TEE_NUM_PARAMS])
+{
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (params[0].value.a != 0 && params[0].value.a != 1)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return rtc_alarm_wake_set_status((bool)params[0].value.a);
+}
+
 static TEE_Result rtc_pta_get_info(uint32_t types,
 				   TEE_Param params[TEE_NUM_PARAMS])
 {
@@ -151,6 +279,10 @@ static TEE_Result rtc_pta_get_info(uint32_t types,
 
 	if (features & RTC_CORRECTION_FEATURE)
 		info->features |= PTA_RTC_FEATURE_CORRECTION;
+	if (features & RTC_ALARM_FEATURE)
+		info->features |= PTA_RTC_FEATURE_ALARM;
+	if (features & RTC_WAKEUP_ALARM)
+		info->features |= PTA_RTC_FEATURE_WAKEUP_ALARM;
 
 	rtc_pta_copy_time_from_optee(&info->range_min, &range_min);
 	rtc_pta_copy_time_from_optee(&info->range_max, &range_max);
@@ -187,6 +319,18 @@ static TEE_Result invoke_command(void *session __unused,
 		return rtc_pta_get_offset(ptypes, params);
 	case PTA_CMD_RTC_SET_OFFSET:
 		return rtc_pta_set_offset(ptypes, params);
+	case PTA_CMD_RTC_READ_ALARM:
+		return rtc_pta_read_alarm(ptypes, params);
+	case PTA_CMD_RTC_SET_ALARM:
+		return rtc_pta_set_alarm(ptypes, params);
+	case PTA_CMD_RTC_ENABLE_ALARM:
+		return rtc_pta_enable_alarm(ptypes, params);
+	case PTA_CMD_RTC_WAIT_ALARM:
+		return rtc_pta_wait_alarm(ptypes, params);
+	case PTA_CMD_RTC_CANCEL_WAIT:
+		return rtc_pta_cancel_wait(ptypes, params);
+	case PTA_CMD_RTC_SET_WAKE_ALARM_STATUS:
+		return rtc_pta_alarm_wake_set_status(ptypes, params);
 	default:
 		break;
 	}
diff --git a/lib/libutee/include/pta_rtc.h b/lib/libutee/include/pta_rtc.h
index ae14c2e1a..72a9da31a 100644
--- a/lib/libutee/include/pta_rtc.h
+++ b/lib/libutee/include/pta_rtc.h
@@ -18,6 +18,21 @@
  */
 #define PTA_RTC_FEATURE_CORRECTION	BIT(0)
 
+/*
+ * RTC provides set/read/enable/wait alarm and thus
+ * commands:
+ * PTA_CMD_RTC_SET_ALARM, PTA_CMD_RTC_READ_ALARM,
+ * PTA_CMD_RTC_WAIT_ALARM, PTA_CMD_RTC_ENABLE_ALARM
+ * might be called
+ */
+#define PTA_RTC_FEATURE_ALARM		BIT(1)
+
+/*
+ * Command PTA_CMD_RTC_SET_WAKE_ALARM_STATUS can be used to enable/disable the
+ * alarm wake-up capability.
+ */
+#define PTA_RTC_FEATURE_WAKEUP_ALARM	BIT(2)
+
 struct pta_rtc_time {
 	uint32_t tm_sec;
 	uint32_t tm_min;
@@ -28,6 +43,18 @@ struct pta_rtc_time {
 	uint32_t tm_wday;
 };
 
+/*
+ * struct pta_rtc_alarm - State of an RTC alarm
+ * @enabled - 1 if alarm is enabled, 0 if disabled
+ * @pending - 1 if alarm event is pending, 0 if not
+ * @time: Alarm elapsure time
+ */
+struct pta_rtc_alarm {
+	uint8_t enabled;
+	uint8_t pending;
+	struct pta_rtc_time time;
+};
+
 /*
  * struct pta_rtc_info - RTC service information
  * @version - 1st 64bit cell, version of the structure: PTA_RTC_INFO_VERSION
@@ -100,4 +127,75 @@ struct pta_rtc_info {
  */
 #define PTA_CMD_RTC_SET_OFFSET		0x4
 
+/*
+ * PTA_CMD_RTC_READ_ALARM - Read RTC alarm
+ *
+ * [out]     memref[0]  RTC buffer memory reference containing a struct
+ *                      pta_rtc_alarm
+ *
+ * Return codes:
+ * TEE_SUCCESS - Invoke command success
+ * TEE_ERROR_BAD_PARAMETERS - Incorrect input param
+ */
+#define PTA_CMD_RTC_READ_ALARM		0x5
+
+/*
+ * PTA_CMD_RTC_SET_ALARM - Set RTC alarm
+ *
+ * [in]     memref[0]  RTC buffer memory reference containing a struct
+ *                     pta_rtc_alarm to be used as RTC alarm
+ *
+ * Return codes:
+ * TEE_SUCCESS - Invoke command success
+ * TEE_ERROR_BAD_PARAMETERS - Incorrect input param
+ */
+#define PTA_CMD_RTC_SET_ALARM		0x6
+
+/*
+ * PTA_CMD_RTC_ENABLE_ALARM - Enable Alarm
+ *
+ * [in]     value[0].a  RTC IRQ flag (uint32_t), 0 to disable the alarm, 1 to
+ *                      enable
+ *
+ * Return codes:
+ * TEE_SUCCESS - Invoke command success
+ * TEE_ERROR_BAD_PARAMETERS - Incorrect input param
+ */
+#define PTA_CMD_RTC_ENABLE_ALARM	0x7
+
+/*
+ * PTA_CMD_RTC_WAIT_ALARM - Get alarm event
+ *
+ * [out]     value[0].a  RTC wait alarm return status (uint32_t):
+ *                       - 0: No alarm event
+ *                       - 1: Alarm event occurred
+ *                       - 2: Alarm event canceled
+ *
+ * Return codes:
+ * TEE_SUCCESS - Invoke command success
+ * TEE_ERROR_BAD_PARAMETERS - Incorrect input param
+ */
+#define PTA_CMD_RTC_WAIT_ALARM		0x8
+
+/*
+ * PTA_CMD_RTC_CANCEL_WAIT - Cancel wait for alarm event
+ *
+ * Return codes:
+ * TEE_SUCCESS - Invoke command success
+ * TEE_ERROR_BAD_PARAMETERS - Incorrect input param
+ */
+#define PTA_CMD_RTC_CANCEL_WAIT		0x9
+
+/*
+ * PTA_CMD_RTC_SET_WAKE_ALARM_STATUS - Set RTC wake alarm status flag
+ *
+ * [in]     value[0].a RTC IRQ wake alarm flag (uint32_t), 0 to disable the wake
+ *                     up capability, 1 to enable.
+ *
+ * Return codes:
+ * TEE_SUCCESS - Invoke command success
+ * TEE_ERROR_BAD_PARAMETERS - Incorrect input param
+ */
+#define PTA_CMD_RTC_SET_WAKE_ALARM_STATUS	0xA
+
 #endif /* __PTA_RTC_H */
-- 
2.47.2

